vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go

```diff
@@ -0,0 +1,166 @@
+package podtolerationsrestriction
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/klog"
+	coreapi "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/serviceaccount"
+)
+
+const (
+	PluginName       = "scheduling.openshift.io/RestrictPodTolerations"
+	masterToleration = "node-role.kubernetes.io/master"
+)
+
+type podTolerations struct {
+	*admission.Handler
+	authorizer authorizer.Authorizer
+}
+
+var _ = initializer.WantsAuthorizer(&podTolerations{})
+var _ = admission.ValidationInterface(&podTolerations{})
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return NewPodTolerationsPlugin(), nil
+	})
+}
+
+func NewPodTolerationsPlugin() admission.Interface {
+	return &podTolerations{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}
+}
+
+func (p *podTolerations) Validate(attr admission.Attributes, _ admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") {
+		return nil
+	}
+	if attr.GetKind().GroupKind() != coreapi.Kind("Pod") {
+		return nil
+	}
+	if len(attr.GetSubresource()) != 0 {
+		return nil
+	}
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected Type %T", attr.GetObject()))
+	}
+	return p.validatePodTolerations(attr, pod.Spec)
+}
+
+func (p *podTolerations) validatePodTolerations(attr admission.Attributes, podSpec coreapi.PodSpec) error {
+	// if a pod has no tolerations, then no need to validate it's tolerations
+	if len(podSpec.Tolerations) == 0 {
+		return nil
+	}
+	tolerations := podSpec.Tolerations
+	if len(podSpec.ServiceAccountName) == 0 {
+		return admission.NewForbidden(attr, fmt.Errorf("serviceaccounts are required on pods that want to use tolerations"))
+	}
+	for _, toleration := range tolerations {
+		allow := p.IsAuthorized(podSpec.ServiceAccountName, attr, toleration)
+		var errs []error
+		if !allow {
+			errs = append(errs, fmt.Errorf("serviceaccounts doesn't have permission to provide the given toleration %v on the pod", toleration.Key))
+		}
+		if len(errs) > 0 {
+			err := v1helpers.NewMultiLineAggregate(errs)
+			return admission.NewForbidden(attr, err)
+		}
+	}
+	return nil
+}
+
+func (p *podTolerations) IsAuthorized(podSAName string, attr admission.Attributes, toleration coreapi.Toleration) bool {
+	saInfo := serviceaccount.UserInfo(attr.GetNamespace(), podSAName, "")
+	// check for hard-coded serviceaccount groups
+	SAInNamespace := serviceaccount.UserInfo(attr.GetNamespace(), "system:serviceaccounts:"+attr.GetNamespace(), "")
```

Comment:
this line and the one below it aren't necessary.  your `saInfo` is fully correct.

Reply:
You mean, we just need saInfo? I thought you wanted hardcoded SA'es to be used as well.

Reply:
> You mean, we just need saInfo? I thought you wanted hardcoded SA'es to be used as well.

I clicked through. that method builds them out for you already.

Reply:
this comment is still outstanding.